### PR TITLE
[ROCm] Implement forward AD for miopen_batch_norm

### PR DESCRIPTION
**Pull request by alugorey**
Implements forward automatic differentiation support for miopen_batch_norm as well as unskips the associated unit tests. Also fixes a class of functorch related unit tests that fail due to failing a contiguous tensor assertion in BatchNorm_miopen.cpp. Solution was to just limit tensors to miopen_batch_norm that have at least 3 dimensions. The exact restriction already existed in the cudnn path and is why the tests in question only failed on ROCm.

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @dllehr-amd @jataylo @hongxiayang

### CODE REVIEW

**Comment by alugorey** [comment]
Hi @zou3519 
This PR addresses several ROCm only functorch failures. Many of which are referenced in this issue: https://github.com/pytorch/pytorch/issues/96560

Will push up a follow up PR addressing the remaining failing tests from the aforementioned issue.

**Review by jeffdaily — APPROVED** [review]

**Comment by andrewor14** [comment]
Hi @alugorey, thanks for fixing this. It seems there are two separate issues, one is forward AD and the other is the contiguous tensor assertion. Is it possible to separate these two? For forward AD, we are planning to migrate to the new batch norm stack (https://github.com/pytorch/pytorch/pull/119496) in the near future so there won't be an `miopen_batch_norm` op anymore. We definitely still want the contiguous tensor change.

**Comment by alugorey** [comment]
> Hi @alugorey, thanks for fixing this. It seems there are two separate issues, one is forward AD and the other is the contiguous tensor assertion. Is it possible to separate these two? For forward AD, we are planning to migrate to the new batch norm stack (#119496) in the near future so there won't be an `miopen_batch_norm` op anymore. We definitely still want the contiguous tensor change.

Hi! Sure. I will roll back the contiguous tensor change from this PR and push up a follow up PR with that change so we can move forward with this one.

**Comment by alugorey** [comment]
@andrewor14 Actually, on closer inspection the reason i wrapped up both those issues into one is because they both fell under the umbrella of the same skip decorator in test_ops.py. https://github.com/pytorch/pytorch/pull/125069/files#diff-ed4bcb4e4dd5f94da948c5505bc39ce2f437fb70af4aa460ea6375f094ff1716L352

Both changes are needed in order to remove this instance of `skipIfRocm` as this decorator unskips multiple tests, some failing due to lack of forwardAD and others due to the contiguous assertion.

**Comment by andrewor14** [comment]
> @andrewor14 Actually, on closer inspection the reason i wrapped up both those issues into one is because they both fell under the umbrella of the same skip decorator in test_ops.py. https://github.com/pytorch/pytorch/pull/125069/files#diff-ed4bcb4e4dd5f94da948c5505bc39ce2f437fb70af4aa460ea6375f094ff1716L352
> 
> Both changes are needed in order to remove this instance of `skipIfRocm` as this decorator unskips multiple tests, some failing due to lack of forwardAD and others due to the contiguous assertion.
> 
> Can we just push this through since the contiguous fix is just a single line and can trivially be brought in? Otherwise we are going to create dependencies.

Hi @alugorey, in that case do you mind waiting for https://github.com/pytorch/pytorch/pull/119496 first? I don't think we need to add forward AD specifically for `miopen_batch_norm` once that PR is merged. It should be handled in the new batch norm ops automatically.

**Comment by alugorey** [comment]
> > @andrewor14 Actually, on closer inspection the reason i wrapped up both those issues into one is because they both fell under the umbrella of the same skip decorator in test_ops.py. https://github.com/pytorch/pytorch/pull/125069/files#diff-ed4bcb4e4dd5f94da948c5505bc39ce2f437fb70af4aa460ea6375f094ff1716L352
> > Both changes are needed in order to remove this instance of `skipIfRocm` as this decorator unskips multiple tests, some failing due to lack of forwardAD and others due to the contiguous assertion.
> > Can we just push this through since the contiguous fix is just a single line and can trivially be brought in? Otherwise we are going to create dependencies.
> 
> Hi @alugorey, in that case do you mind waiting for #119496 first? I don't think we need to add forward AD specifically for `miopen_batch_norm` once that PR is merged. It should be handled in the new batch norm ops automatically.

My concern is the unit tests that this PR enables. I think your solution needs to be able to pass everything that now passes as a result of this PR. We won't know that unless we push this through and enable those tests. There seems to be minimal overlap between my PR and this new one so a simple rebase should suffice and we would get the benefit of making sure the new stuff passes even more tests. I don't see the harm in pushing this one through.

@jeffdaily @albanD @zou3519 @jithunnair-amd  thoughts here?

**Comment by andrewor14** [comment]
My concern is that the forward AD changes here will just become dead code in the future, since it's not aligned with our long term plans of consolidating batch norm ops (and removing `aten.miopen_batch_norm`). If this is a change needed urgently before #119496 lands (possibly by the end of the month) then maybe we can go forward with this PR. Otherwise I feel it's better to just let #119496 handle the forward AD part generally. I'll let @albanD decide.

**Comment by albanD** [comment]
I think we can just merge this, the reason is:
- We don't have a strict timeline for when https://github.com/pytorch/pytorch/pull/119496 is going to land right now.
- Remove test skips is always good, this will add more signal for our PR to make sure everything is working smoothly.
- This doesn't add significant work on top of the cleanup work we are already planning on doing post-https://github.com/pytorch/pytorch/pull/119496
- Everyone here understands that this code will be deleted within the next 6months and so this is strictly a temporary thing.

**Comment by jeffdaily** [comment]
@andrewor14 or @albanD, since it sounds decided, can we get an approval so we can merge?

**Review by andrewor14 — APPROVED** [review]
